### PR TITLE
fix: add controller to as-vending

### LINKED ISSUE #41: add controller to as-vending

### PR DESCRIPTION
**Pull request by joshua-silverio**
Closes #41 
Signed-off-by: intel <joshua.silverio@intel.com>

### CODE REVIEW

File: as-vending/routes/controller.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+	utilities.ProcessCORS(writer, req, func(writer http.ResponseWriter, req *http.Request) {
+		mm, _ := utilities.GetAsJSON(functions.MaintenanceMode{MaintenanceMode: c.vendingState.MaintenanceMode})
+		utilities.WriteJSONHTTPResponse(writer, req, http.StatusOK, mm, false)
+	})
+}
+
+func (c *Controller) errorAddRouteHandler(err error) error {
+	if err != nil {
+		c.lc.Errorf("error adding route: %s", err.Error())
+		return fmt.Errorf("error adding route: %s", err.Error())
```
Comment (jim-wang-intel) [inline as-vending/routes/controller.go:70]: could use error msg variable to reduce the duplicate hard code err message

File: as-vending/routes/controller_test.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	t.Run("TestGetMaintenanceMode MaintenanceMode=True", func(t *testing.T) {
+		var vendingState functions.VendingState
+		var maintModeAPIResponse functions.MaintenanceMode
+		c := NewController(logger.NewMockClient(), nil, vendingState)
+		// set the vendingState's MaintenanceMode boolean accordingly
+		c.vendingState.MaintenanceMode = true
+
+		req := httptest.NewRequest("GET", "/maintenanceMode", nil)
```
Comment (jim-wang-intel) [inline as-vending/routes/controller_test.go:87]: http.MethodGet

File: as-vending/routes/controller_test.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+		assert.Equal(t, maintModeAPIResponse, maintModeTrue, "Received a maintenance mode response that was different than anticipated")
+	})
+	t.Run("TestGetMaintenanceMode MaintenanceMode=False", func(t *testing.T) {
+		var vendingState functions.VendingState
+		var maintModeAPIResponse functions.MaintenanceMode
+		c := NewController(logger.NewMockClient(), nil, vendingState)
+		// set the vendingState's MaintenanceMode boolean accordingly
+		c.vendingState.MaintenanceMode = false
+
+		req := httptest.NewRequest("GET", "/maintenanceMode", nil)
```
Comment (jim-wang-intel) [inline as-vending/routes/controller_test.go:108]: http.MethodGet

File: as-vending/routes/controller_test.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+		assert.Equal(t, resp.Status, "200 OK", "OPTIONS request did not return 200")
+	})
+}
+
+func TestResetDoorLock(t *testing.T) {
+	stopChannel := make(chan int)
+	var vendingState functions.VendingState
+	vendingState.ThreadStopChannel = stopChannel
+	c := NewController(logger.NewMockClient(), nil, vendingState)
+	request, _ := http.NewRequest("POST", "", nil)
```
Comment (jim-wang-intel) [inline as-vending/routes/controller_test.go:145]: http.MethodPost

File: as-vending/routes/controller_test.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	}
+	boardStatus := functions.ControllerBoardStatus{
+		MaxTemperatureStatus: true,
+		MinTemperatureStatus: true,
+		DoorClosed:           true,
+	}
+
+	b, _ := json.Marshal(boardStatus)
+	c := NewController(logger.NewMockClient(), nil, vendingState)
+	request, _ := http.NewRequest("POST", "", bytes.NewBuffer(b))
```
Comment (jim-wang-intel) [inline as-vending/routes/controller_test.go:178]: http.MethodPost

File: as-vending/routes/controller.go (hunk 1 above)
Comment (joshua-silverio) [reply as-vending/routes/controller.go]: updated

File: as-vending/routes/controller_test.go (hunk 2 above)
Comment (joshua-silverio) [reply as-vending/routes/controller_test.go]: fixed

File: as-vending/main.go [hunk 6]
```diff
@@ -73,21 +73,16 @@ func main() {
...
+	if err != nil {
+		lc.Errorf("failed to add all Routes: %s", err.Error())
+		os.Exit(1)
+	}
 	// create the function pipeline to run when an event is read on the device channels
 	err = service.SetFunctionsPipeline(
-		transforms.NewFilterFor(vendingState.Configuration.DeviceNames).FilterByDeviceName,
-		vendingState.DeviceHelper,
+		transforms.NewFilterFor(controller.GetVendingState().Configuration.DeviceNames).FilterByDeviceName,
+		controller.GetVendingState().DeviceHelper,
```
Comment (lenny-intel) [inline as-vending/main.go:85]: This Pipeline Function doesn't belong in the HTTP Controller.

File: as-vending/main.go [hunk 7]
```diff
@@ -73,21 +73,16 @@ func main() {
...
+	err := controller.AddAllRoutes()
+	if err != nil {
+		lc.Errorf("failed to add all Routes: %s", err.Error())
+		os.Exit(1)
+	}
 	// create the function pipeline to run when an event is read on the device channels
 	err = service.SetFunctionsPipeline(
-		transforms.NewFilterFor(vendingState.Configuration.DeviceNames).FilterByDeviceName,
-		vendingState.DeviceHelper,
+		transforms.NewFilterFor(controller.GetVendingState().Configuration.DeviceNames).FilterByDeviceName,
```
Comment (lenny-intel) [inline as-vending/main.go:84]: `GetVendingState` doesn't seem appropriate on the Controller.

**Review by lenny-intel — CHANGES_REQUESTED** [review]
Got more to review, but ran out of time. So here is what I have so far.

File: as-vending/main.go (hunk 7 above)
Comment (joshua-silverio) [reply as-vending/main.go:84]: other option is to expose vending state in controller struct?

File: as-vending/routes/controller_test.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/edgexfoundry/app-functions-sdk-go/v2/pkg/interfaces/mocks"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/clients/logger"
+	utilities "github.com/intel-iot-devkit/automated-checkout-utilities"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"gotest.tools/assert"
```
[the remark this answers is not in the document]
Comment (joshua-silverio) [reply as-vending/routes/controller_test.go]: changed

**Review by jim-wang-intel — APPROVED** [review]
LGTM

File: as-vending/main.go (hunk 7 above)
Comment (vli11) [reply as-vending/main.go:84]: vending state is already in main for you to use

Comment (brian-intel) [reply as-vending/main.go:84]: This is the same issue we are seeing in as-controller-board. VendingState is essentially our controller but a different name. One option is to move the VendingState struct as a variable of the controller and change how we are getting the data.

**Review by lenny-intel — APPROVED** [review]
Nice Job!!

**Pull request merged**